tests/core/test_utilities.py

```diff
@@ -913,6 +914,81 @@ def test_fit_plane_to_points():
     )
 
 
+DEFAULT_PRINCIPAL_AXES = [[0.0, 0.0, 1.0], [0.0, 1.0, 0.0], [-1.0, 0.0, 0.0]]
+
+CASE_RANK0 = (
+    0,
+    [[0, 0, 0], [0, 0, 0]],
+    [DEFAULT_PRINCIPAL_AXES],
+)
+CASE_RANK1 = (
+    1,
+    [[0, 0, 0], [1, 0, 0]],
+    [[1.0, 0.0, 0.0], [0.0, 0.0, 1.0], [0.0, -1.0, 0.0]],
+)
+
+CASE_RANK2 = (
+    2,
+    [[1, 2, 3], [3, 2, 1]],
+    [
+        [-0.70710678, 0.0, 0.70710678],
+        [0.0, 1.0, 0.0],
+        [-0.70710678, 0.0, -0.70710678],
+    ],
+)
+CASE_RANK3 = (
+    3,
+    pv.examples.load_airplane().points,
+    [
+        [8.1159601e-07, -9.9255711e-01, -1.2178054e-01],
+        [-1.0000000e00, -8.0911440e-07, -6.9828459e-08],
+        [-2.9225653e-08, 1.2178054e-01, -9.9255711e-01],
+    ],
+)
+
+
+@pytest.mark.parametrize(
+    ('rank', 'points', 'expected_axes'),
+    [CASE_RANK0, CASE_RANK1, CASE_RANK2, CASE_RANK3],
+    ids=['rank0', 'rank1', 'rank2', 'rank3'],
+)
+def test_principal_axes(rank, points, expected_axes):
+    assert np.linalg.matrix_rank(points) == rank
+
+    axes = principal_axes(points)
+    assert np.allclose(axes, expected_axes, atol=1e-7)
+
+    assert np.array_equal(np.cross(axes[0], axes[1]), axes[2])
+    assert np.allclose(np.linalg.norm(axes[0]), 1)
+    assert np.allclose(np.linalg.norm(axes[1]), 1)
+    assert np.allclose(np.linalg.norm(axes[2]), 1)
+
+    assert type(axes) is np.ndarray
+
+    _, sizes = principal_axes(points, return_sizes=True)
+    assert sizes[0] >= sizes[1]
+    assert sizes[1] >= sizes[2]
+    assert type(sizes) is np.ndarray
+
+
+def test_principal_axes_empty():
+    axes = principal_axes(np.empty((0, 3)))
+    assert np.allclose(axes, DEFAULT_PRINCIPAL_AXES)
+
+
+def test_principal_axes_single_point():
+    axes = principal_axes([1, 2, 3])
+    assert np.allclose(axes, DEFAULT_PRINCIPAL_AXES)
+
+
+def test_principal_axes_vectors_many_points():
+    n_points = 2_000_000
+    points = np.random.default_rng().random((n_points, 3))
+    axes = pv.principal_axes(points)
+    assert np.any(axes)
+    assert np.all(axes != 0)
```

Comment:
1. What does this test exactly? Just that there's no error for "many" points? Note that `np.all(axes != 0)` implies `np.any(axes)`. And for any kind of random float data `!= 0` will almost certainly be true. This is why I'm not sure what you're confirming here.
2. Mostly a style note, but since we ensure we have `ndarray`s here, I'd go with `axes.any()`/`axes.all()` instance methods.

Reply:
> 1. What does this test exactly? Just that there's no error for "many" points? 

Yes, this is because a similar test using `np.linalg.svd` instead of the PCA method fails with numpy raising array memory errors. So I wanted to make sure this is no longer the case.

> Note that `np.all(axes != 0)` implies `np.any(axes)`. And for any kind of random float data `!= 0` will almost certainly be true. This is why I'm not sure what you're confirming here.

This part isn't really necessary, but I was intending to make sure that the output wasn't simply the default output, i.e. checking that the computation didn't just silently fail and return zeros and ones.

Reply:
Hmm, then perhaps you could use this test to check the `size`s? Not sure if it's an issue to mix test semantics. You could use a dense ellipsoid with "many" points and check the resulting sizes while you're at it. Might be a bad idea, I'm not sure.

Reply:
It's a good idea, I like this better. I've updated the test to make the intent more clear, and use a high-resolution ellipsoid and check the returned sizes.